File: src/m/test_trapezoidal.m
function test_trapezoidal_one_exact_result
  % Compare one hand-computed result
  v = @(t) 3*(t^2)*exp(t^3);
  n = 2;
  numerical = trapezoidal(v, 0, 1, n);
  exact = 2.463642041244344;
  err = abs(exact - numerical);
  tol = 1E-14;
  assert(err < tol, 'error=%g > tol=%g', err, tol);
end

function test_trapezoidal_linear
  % Check that linear functions are integrated exactly
  f = @(x) 6*x - 4;
  F = @(x) 3*x**2 - 4*x;  % Anti-derivative
  a = 1.2; b = 4.4;
  exact = F(b) - F(a);
  tol = 1E-14;
  for n = [2 20 21]
    numerical = trapezoidal(f, a, b, n);
    err = abs(exact - numerical);
    assert(err < tol, 'n=%d, err=%g', n, err);
  end
end

function r = convergence_rates(f, F, a, b, num_experiments)
  n = zeros(num_experiments, 1);
  E = zeros(num_experiments, 1);
  r = zeros(num_experiments-1, 1);
  exact = F(b) - F(a);

  for i = 1:num_experiments
    n(i) = 2^i;
    numerical = trapezoidal(f, a, b, n(i));
    err = abs(exact - numerical);
    E(i) = err;
    if (i > 1)
      r(i-1) = log(E(i-1)/E(i))/log(n(i-1)/n(i));
      r(i-1) = round(r(i-1)*100)/100;  % Truncate, two decimals
   end
  end
end

function test_trapezoidal_conv_rate
  % Check empirical convergence rates against the expected -2.
  v = @(t) 3*(t^2)*exp(t^3);
  V = @(t) exp(t^3);
  a = 1.1; b = 1.9;
  num_experiments = 14;
  r = convergence_rates(v, V, a, b, num_experiments)
  tol = 0.01;
  assert(abs(r(num_experiments-1)) - 2 < tol, '%f, %f, %f, %f, %f',...
                   r((num_experiments-1)-4:num_experiments-1));
end

test_trapezoidal_one_exact_result()
test_trapezoidal_linear()
test_trapezoidal_conv_rate()

File: src/py/test_trapezoidal.py
from trapezoidal import trapezoidal

def test_trapezoidal_one_exact_result():
    """Compare one hand-computed result."""
    from math import exp
    v = lambda t: 3*(t**2)*exp(t**3)
    n = 2
    computed = trapezoidal(v, 0, 1, n)
    expected = 2.463642041244344
    error = abs(expected - computed)
    tol = 1E-14
    success = error < tol
    msg = 'error=%g > tol=%g' % (error, tol)
    assert success, msg

def test_trapezoidal_linear():
    """Check that linear functions are integrated exactly."""
    f = lambda x: 6*x - 4
    F = lambda x: 3*x**2 - 4*x  # Anti-derivative
    a = 1.2; b = 4.4
    expected = F(b) - F(a)
    tol = 1E-14
    for n in 2, 20, 21:
        computed = trapezoidal(f, a, b, n)
        error = abs(expected - computed)
        success = error < tol
        msg = 'n=%d, err=%g' % (n, error)
        assert success, msg

def convergence_rates(f, F, a, b, num_experiments=14):
    from math import log
    from numpy import zeros
    expected = F(b) - F(a)
    n = zeros(num_experiments, dtype=int)
    E = zeros(num_experiments)
    r = zeros(num_experiments-1)
    for i in range(num_experiments):
        n[i] = 2**(i+1)
        computed = trapezoidal(f, a, b, n[i])
        E[i] = abs(expected - computed)
        if i > 0:
            r_im1 = log(E[i]/E[i-1])/log(float(n[i])/n[i-1])
            r[i-1] = float('%.2f' % r_im1) # Truncate to two decimals
    return r

def test_trapezoidal_conv_rate():
    """Check empirical convergence rates against the expected -2."""
    from math import exp
    v = lambda t: 3*(t**2)*exp(t**3)
    V = lambda t: exp(t**3)
    a = 1.1; b = 1.9
    r = convergence_rates(v, V, a, b, 14)
    print r
    tol = 0.01
    msg = str(r[-4:])  # show last 4 estimated rates
    assert (abs(r[-1]) - 2) < tol, msg
